doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,796 @@
+using special methods on the datatypes to implement any necessary logic.
+This NEP suggests a specific choice of API, and set of these methods.
+Here these are suggested as C-API slots, however, this translates identically
+to python methods.
+
+Additionally, the NEP proposes to implement the notion of *abstract* dtypes
+below.
+Further, we detail – in part – how the proposed methods (C-API slots)
+enable all necessary use cases.
+
+Each section below will begin with a short description of the reason for the
+specific choice.
+
+
+Nomenclature
+^^^^^^^^^^^^
+
+As a brief note on nomenclature, it should be noted that ``dtype`` normally
+denotes the dtype *instance*, which is the object attached to a numpy array.
+On the other hand the ``DType`` class or type is the subclass of ``np.dtype``.
+On the C-level we currently use the name ``descriptor`` or ``descr`` in many
+cases interchangeable with the dtype instance, this will be used in proposed
+C-API names for clarification.
+Note that the notion of dtype class is currently represented only as
+the ``dtype.num`` and ``dtype.char``.
+Please see the dtype hierarchy figure as an illustration of this important
+distinction.
+
+Currently classes do exist as ``np.float64``, however,
+these are not dtypes but the corresponding scalar classes
+(see also NEP 40 and 41 on a discussion why these are largely unrelated to the proposed changes).
+
+
+Hierarchy of DataTypes and Abstract DTypes
+""""""""""""""""""""""""""""""""""""""""""
+
+**Motivation:**
+The creation of a DType classes has already been decided in NEP 41.
+This proposes the notion of _abstract_ DTypes.
+There are multiple reasons for this:
+1. It allows the definition of a class hierarchy, allowing in principle to write
+   ``isinstance(np.dtype("float64"), np.Inexact)``.
+   **DEPENDING ON WHAT WE DECIDE FOR UFUNCS, THAT MAY BE IMPORTANT THERE**
+2. Allowing to write such an abstract datatype can enable code such as
+   ``arr.astype(np.Complex)`` as a way to spell the desire to cast to a
+   complex data type.
+3. It anticipates the creation of families of DTypes by users.
+   For example allowing the creation of a ``Unit`` with a concrete
+   ``Float64Unit``.
+
+
+**Description:**
+The figure below shows the proposed datatype hierarchy.
+It should be noted that abstract DTypes are distinct mainly in two ways:
+
+1. They do not have instances. Instantiating an abstract DType has to return
+   a concrete subclass.
+2. Unlike concrete DTypes, abstract DTypes can be superclasses, they may also
+   serve like pythons abstract base classes (ABC).
+
+These two rules are identical to the type choices made for example in the Julia
+language.
+It allows for the creation of a datatype hierarchy, but avoids issues with
+subclassing concrete DTypes directly:
+Logic such as ``can_cast`` does not cleanly inherit e.g. from a
+``Int64`` to a ``Datetime64`` even though the ``Datetime64`` could be seen
```

Comment:
Yes, each datetime64 with different unit is its own instance. It is a good example to use throughout, since units are one of the most complex use-cases. I think *unsafe* casting is OK for these (i.e. it is allowed to lose the unit).

Reply:
Great to use it!

Though personally I do not think it should be OK to cast unsafely - you have not just thrown away precision or so, but the actual meaning of the number - just like you would if you cast from meters to seconds - I think this is something you should use  `.view(integer)` for. (In astropy, for any number with units we allow only 0, infinity, and nan to have arbitrary unit - which does break down for 0 for Celsius, but then, that is a bad unit!)

Reply:
That is a fair point, maybe we actually need an "even more unsafe" casting :). Note that precision loss in some sense covered by `"same_kind"` casting? I had not thought about it too much that our current "unsafe" casting. However, I think you could also do this:

* `UnitFloat64[m] -> Float64` (cannot cast!)
* `UnitFloat64[unitless] -> Float64` unsafe cast.

Now the question is, how do you get from `UnitFloat64[m]` to `UnitFloat64[unitless]`. We could even decide that you need a UFunc to do that. Or that step is "unsafe" (and a view), and you simply need two steps to get to the actual `float64`. Requiring two steps seems completely fine to me, but it seems like we may still need another level of safeness, basically what you call `view` which currently means: ignore casting safety entirely.

Reply:
Your itemized example is how astropy works! But you are right that of course our quantities do have a `.value` property that just gives the unit-less version (it is just a view), and one uses this quite a bit (as one does the `to_value(new-unit)` which gives the value in the new unit). So, it may have to be that the dtype itself introduces a new casting rule? Maybe those casting strings can become dtype-specific? For now, probably best to leave open!!

Reply:
Yes, we could specify (for new slots) that they have to allow for new casting rules to be added in the future. But I actually think that if you are concerned that "unsafe" is not enough, the custom ufunc is not unreasonable: If you make unit, add `units.force_unit(arr, dtype=units.Float64("m"))` ufunc.